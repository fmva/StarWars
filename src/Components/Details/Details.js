import { Component } from 'react';
import React from 'react';
import { Button, Modal } from 'antd';
import './Details.less';
import PropTypes from 'prop-types';
import Resources from '../../Resources/Resources';

class Details extends Component {
  static propTypes = {
    /** array of details */
    details: PropTypes.array,
    /** close modal handler */
    onClose: PropTypes.func
  };

  /**
   * render detail list
   * @param {Array} arr - array of details
   */
  renderDetailList(arr) {
    const listDetail = arr.map((item, key) => (
      <li key={`details_${key}}`} className="detail-row">
        {item.map((itemInner, keyInner) => (
          <div key={`details_column_${keyInner}}`} className="detail-column">
            <span className="detail-title">{itemInner.title}</span>
            <span className="detail-value">{itemInner.value}</span>
          </div>
        ))}
      </li>
    ));
    return <ul className="detail-main">{listDetail}</ul>;
  }

  render() {
    const details = this.props.details && this.props.details.length > 0 ? this.props.details : [];
    const onClose = this.props.onClose ? this.props.onClose : [];
    return (
      <div className="modal-block">
        <Modal
          title={Resources.getResource('details_title')}
          visible={details.length > 0}
          onCancel={onClose}
          footer={[
            <Button key="back" type="primary" onClick={onClose}>
              {Resources.getResource('ok_button')}
            </Button>
          ]}
        >
          {this.renderDetailList(details)}
        </Modal>
      </div>
    );
  }
}

export default Details;
