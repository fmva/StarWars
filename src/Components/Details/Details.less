@import "../../Less/overall.less";

ul.detail-main {
  list-style-type: none;
  margin:0;
  padding:0;
  color: #000;
  font-size:1.3em;
}

.detail-row {
  display: -webkit-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  padding: 4px 0;
  @media @phone {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    padding: 0;
  }
}

.detail-column{
  width:50%;
  @media @phone {
    width: auto;
  }
}

.detail-row div {
  padding:0 5px;
}

.detail-column span{
  padding: 2px;
}

.detail-title{
  font-weight: bold;
}