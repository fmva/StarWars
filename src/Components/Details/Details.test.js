import Details from './Details';
import { shallow, mount } from 'enzyme';
import React from 'react';

const details = [
  [{ title: 'title 1', value: 'value 1' }, { title: 'title 2', value: 'value 2' }],
  [{ title: 'title 3', value: 'value 3' }, { title: 'title 4', value: 'value 4' }]
];
const mockOnClose = jest.fn();

describe('Details', () => {
  it('Компонент Details с входными данными по какой-либо сущности', () => {
    const wrapper = shallow(<Details details={details} onClose={mockOnClose} />);
    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find('.detail-main li').length).toBe(2);
    expect(wrapper.find('.detail-title').length).toBe(4);

    expect(
      wrapper
        .find('.detail-title')
        .at(0)
        .text()
    ).toBe('title 1');
    expect(
      wrapper
        .find('.detail-title')
        .at(1)
        .text()
    ).toBe('title 2');
    expect(
      wrapper
        .find('.detail-title')
        .at(2)
        .text()
    ).toBe('title 3');
    expect(
      wrapper
        .find('.detail-title')
        .at(3)
        .text()
    ).toBe('title 4');

    expect(wrapper.find('.detail-value').length).toBe(4);
    expect(
      wrapper
        .find('.detail-value')
        .at(0)
        .text()
    ).toBe('value 1');
    expect(
      wrapper
        .find('.detail-value')
        .at(1)
        .text()
    ).toBe('value 2');
    expect(
      wrapper
        .find('.detail-value')
        .at(2)
        .text()
    ).toBe('value 3');
    expect(
      wrapper
        .find('.detail-value')
        .at(3)
        .text()
    ).toBe('value 4');
  });
});

it('При клике на кнопку ОК должно сработать событие onClose', () => {
  const wrapper = mount(<Details details={details} onClose={mockOnClose} />);
  expect(wrapper).toMatchSnapshot();

  wrapper.find('button.ant-modal-close').simulate('click');
  expect(mockOnClose.mock.calls.length).toBe(1);
});
