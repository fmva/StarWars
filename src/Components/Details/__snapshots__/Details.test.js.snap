// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Компонент Details с входными данными по какой-либо сущности 1`] = `
<div
  className="modal-block"
>
  <Modal
    cancelButtonDisabled={false}
    confirmLoading={false}
    footer={
      Array [
        <Button
          block={false}
          ghost={false}
          loading={false}
          onClick={[MockFunction]}
          type="primary"
        >
          OK
        </Button>,
      ]
    }
    maskTransitionName="fade"
    okButtonDisabled={false}
    okType="primary"
    onCancel={[MockFunction]}
    title="Details"
    transitionName="zoom"
    visible={true}
    width={520}
  >
    <ul
      className="detail-main"
    >
      <li
        className="detail-row"
        key="details_0}"
      >
        <div
          className="detail-column"
          key="details_column_0}"
        >
          <span
            className="detail-title"
          >
            title 1
          </span>
          <span
            className="detail-value"
          >
            value 1
          </span>
        </div>
        <div
          className="detail-column"
          key="details_column_1}"
        >
          <span
            className="detail-title"
          >
            title 2
          </span>
          <span
            className="detail-value"
          >
            value 2
          </span>
        </div>
      </li>
      <li
        className="detail-row"
        key="details_1}"
      >
        <div
          className="detail-column"
          key="details_column_0}"
        >
          <span
            className="detail-title"
          >
            title 3
          </span>
          <span
            className="detail-value"
          >
            value 3
          </span>
        </div>
        <div
          className="detail-column"
          key="details_column_1}"
        >
          <span
            className="detail-title"
          >
            title 4
          </span>
          <span
            className="detail-value"
          >
            value 4
          </span>
        </div>
      </li>
    </ul>
  </Modal>
</div>
`;

exports[`При клике на кнопку ОК должно сработать событие onClose 1`] = `
<Details
  details={
    Array [
      Array [
        Object {
          "title": "title 1",
          "value": "value 1",
        },
        Object {
          "title": "title 2",
          "value": "value 2",
        },
      ],
      Array [
        Object {
          "title": "title 3",
          "value": "value 3",
        },
        Object {
          "title": "title 4",
          "value": "value 4",
        },
      ],
    ]
  }
  onClose={[MockFunction]}
>
  <div
    className="modal-block"
  >
    <Modal
      cancelButtonDisabled={false}
      confirmLoading={false}
      footer={
        Array [
          <Button
            block={false}
            ghost={false}
            loading={false}
            onClick={[MockFunction]}
            type="primary"
          >
            OK
          </Button>,
        ]
      }
      maskTransitionName="fade"
      okButtonDisabled={false}
      okType="primary"
      onCancel={[MockFunction]}
      title="Details"
      transitionName="zoom"
      visible={true}
      width={520}
    >
      <DialogWrap
        cancelButtonDisabled={false}
        closeIcon={
          <span
            className="ant-modal-close-x"
          >
            <Icon
              className="ant-modal-close-icon"
              type="close"
            />
          </span>
        }
        confirmLoading={false}
        footer={
          Array [
            <Button
              block={false}
              ghost={false}
              loading={false}
              onClick={[MockFunction]}
              type="primary"
            >
              OK
            </Button>,
          ]
        }
        forceRender={false}
        maskTransitionName="fade"
        okButtonDisabled={false}
        okType="primary"
        onCancel={[MockFunction]}
        onClose={[Function]}
        prefixCls="ant-modal"
        title="Details"
        transitionName="zoom"
        visible={true}
        width={520}
        wrapClassName=""
      >
        <Portal
          getContainer={[Function]}
        >
          <Portal
            containerInfo={
              <div>
                <div>
                  <div
                    class="ant-modal-mask fade-appear"
                  />
                  <div
                    aria-labelledby="rcDialogTitle0"
                    class="ant-modal-wrap "
                    role="dialog"
                    tabindex="-1"
                  >
                    <div
                      class="ant-modal zoom-appear"
                      role="document"
                      style="width: 520px;"
                    >
                      <div
                        style="width: 0px; height: 0px; overflow: hidden;"
                        tabindex="0"
                      >
                        sentinelStart
                      </div>
                      <div
                        class="ant-modal-content"
                      >
                        <button
                          aria-label="Close"
                          class="ant-modal-close"
                        >
                          <span
                            class="ant-modal-close-x"
                          >
                            <i
                              aria-label="icon: close"
                              class="anticon anticon-close ant-modal-close-icon"
                            >
                              <svg
                                aria-hidden="true"
                                class=""
                                data-icon="close"
                                fill="currentColor"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                />
                              </svg>
                            </i>
                          </span>
                        </button>
                        <div
                          class="ant-modal-header"
                        >
                          <div
                            class="ant-modal-title"
                            id="rcDialogTitle0"
                          >
                            Details
                          </div>
                        </div>
                        <div
                          class="ant-modal-body"
                        >
                          <ul
                            class="detail-main"
                          >
                            <li
                              class="detail-row"
                            >
                              <div
                                class="detail-column"
                              >
                                <span
                                  class="detail-title"
                                >
                                  title 1
                                </span>
                                <span
                                  class="detail-value"
                                >
                                  value 1
                                </span>
                              </div>
                              <div
                                class="detail-column"
                              >
                                <span
                                  class="detail-title"
                                >
                                  title 2
                                </span>
                                <span
                                  class="detail-value"
                                >
                                  value 2
                                </span>
                              </div>
                            </li>
                            <li
                              class="detail-row"
                            >
                              <div
                                class="detail-column"
                              >
                                <span
                                  class="detail-title"
                                >
                                  title 3
                                </span>
                                <span
                                  class="detail-value"
                                >
                                  value 3
                                </span>
                              </div>
                              <div
                                class="detail-column"
                              >
                                <span
                                  class="detail-title"
                                >
                                  title 4
                                </span>
                                <span
                                  class="detail-value"
                                >
                                  value 4
                                </span>
                              </div>
                            </li>
                          </ul>
                        </div>
                        <div
                          class="ant-modal-footer"
                        >
                          <button
                            class="ant-btn ant-btn-primary"
                            type="button"
                          >
                            <span>
                              OK
                            </span>
                          </button>
                        </div>
                      </div>
                      <div
                        style="width: 0px; height: 0px; overflow: hidden;"
                        tabindex="0"
                      >
                        sentinelEnd
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            }
          >
            <Dialog
              cancelButtonDisabled={false}
              className=""
              closable={true}
              closeIcon={
                <span
                  className="ant-modal-close-x"
                >
                  <Icon
                    className="ant-modal-close-icon"
                    type="close"
                  />
                </span>
              }
              confirmLoading={false}
              destroyOnClose={false}
              footer={
                Array [
                  <Button
                    block={false}
                    ghost={false}
                    loading={false}
                    onClick={[MockFunction]}
                    type="primary"
                  >
                    OK
                  </Button>,
                ]
              }
              forceRender={false}
              key="dialog"
              keyboard={true}
              mask={true}
              maskClosable={true}
              maskTransitionName="fade"
              okButtonDisabled={false}
              okType="primary"
              onCancel={[MockFunction]}
              onClose={[Function]}
              prefixCls="ant-modal"
              title="Details"
              transitionName="zoom"
              visible={true}
              width={520}
              wrapClassName=""
            >
              <div>
                <Animate
                  animation={Object {}}
                  component=""
                  componentProps={Object {}}
                  key="mask"
                  onAppear={[Function]}
                  onEnd={[Function]}
                  onEnter={[Function]}
                  onLeave={[Function]}
                  showProp="visible"
                  transitionAppear={true}
                  transitionEnter={true}
                  transitionLeave={true}
                  transitionName="fade"
                >
                  <AnimateChild
                    animation={Object {}}
                    key="mask"
                    transitionAppear={true}
                    transitionEnter={true}
                    transitionLeave={true}
                    transitionName="fade"
                  >
                    <LazyRenderBox
                      className="ant-modal-mask"
                      hiddenClassName="ant-modal-mask-hidden"
                      key="mask"
                      style={Object {}}
                      visible={true}
                    >
                      <div
                        className="ant-modal-mask"
                        style={Object {}}
                      />
                    </LazyRenderBox>
                  </AnimateChild>
                </Animate>
                <div
                  aria-labelledby="rcDialogTitle0"
                  className="ant-modal-wrap "
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="dialog"
                  style={
                    Object {
                      "display": null,
                    }
                  }
                  tabIndex={-1}
                >
                  <Animate
                    animation={Object {}}
                    component=""
                    componentProps={Object {}}
                    key="dialog"
                    onAppear={[Function]}
                    onEnd={[Function]}
                    onEnter={[Function]}
                    onLeave={[Function]}
                    showProp="visible"
                    transitionAppear={true}
                    transitionEnter={true}
                    transitionLeave={true}
                    transitionName="zoom"
                  >
                    <AnimateChild
                      animation={Object {}}
                      key="dialog-element"
                      transitionAppear={true}
                      transitionEnter={true}
                      transitionLeave={true}
                      transitionName="zoom"
                    >
                      <LazyRenderBox
                        className="ant-modal "
                        key="dialog-element"
                        role="document"
                        style={
                          Object {
                            "width": 520,
                          }
                        }
                        visible={true}
                      >
                        <div
                          className="ant-modal "
                          role="document"
                          style={
                            Object {
                              "width": 520,
                            }
                          }
                        >
                          <div
                            style={
                              Object {
                                "height": 0,
                                "overflow": "hidden",
                                "width": 0,
                              }
                            }
                            tabIndex={0}
                          >
                            sentinelStart
                          </div>
                          <div
                            className="ant-modal-content"
                          >
                            <button
                              aria-label="Close"
                              className="ant-modal-close"
                              onClick={[Function]}
                            >
                              <span
                                className="ant-modal-close-x"
                              >
                                <Icon
                                  className="ant-modal-close-icon"
                                  type="close"
                                >
                                  <LocaleReceiver
                                    componentName="Icon"
                                  >
                                    <i
                                      aria-label="icon: close"
                                      className="anticon anticon-close ant-modal-close-icon"
                                    >
                                      <IconReact
                                        className=""
                                        type="close-o"
                                      >
                                        <svg
                                          aria-hidden="true"
                                          className=""
                                          data-icon="close"
                                          fill="currentColor"
                                          height="1em"
                                          key="svg-close"
                                          viewBox="64 64 896 896"
                                          width="1em"
                                        >
                                          <path
                                            d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 0 0 203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                                            key="svg-close-svg-0"
                                          />
                                        </svg>
                                      </IconReact>
                                    </i>
                                  </LocaleReceiver>
                                </Icon>
                              </span>
                            </button>
                            <div
                              className="ant-modal-header"
                            >
                              <div
                                className="ant-modal-title"
                                id="rcDialogTitle0"
                              >
                                Details
                              </div>
                            </div>
                            <div
                              className="ant-modal-body"
                            >
                              <ul
                                className="detail-main"
                              >
                                <li
                                  className="detail-row"
                                  key="details_0}"
                                >
                                  <div
                                    className="detail-column"
                                    key="details_column_0}"
                                  >
                                    <span
                                      className="detail-title"
                                    >
                                      title 1
                                    </span>
                                    <span
                                      className="detail-value"
                                    >
                                      value 1
                                    </span>
                                  </div>
                                  <div
                                    className="detail-column"
                                    key="details_column_1}"
                                  >
                                    <span
                                      className="detail-title"
                                    >
                                      title 2
                                    </span>
                                    <span
                                      className="detail-value"
                                    >
                                      value 2
                                    </span>
                                  </div>
                                </li>
                                <li
                                  className="detail-row"
                                  key="details_1}"
                                >
                                  <div
                                    className="detail-column"
                                    key="details_column_0}"
                                  >
                                    <span
                                      className="detail-title"
                                    >
                                      title 3
                                    </span>
                                    <span
                                      className="detail-value"
                                    >
                                      value 3
                                    </span>
                                  </div>
                                  <div
                                    className="detail-column"
                                    key="details_column_1}"
                                  >
                                    <span
                                      className="detail-title"
                                    >
                                      title 4
                                    </span>
                                    <span
                                      className="detail-value"
                                    >
                                      value 4
                                    </span>
                                  </div>
                                </li>
                              </ul>
                            </div>
                            <div
                              className="ant-modal-footer"
                            >
                              <Button
                                block={false}
                                ghost={false}
                                key="back"
                                loading={false}
                                onClick={[MockFunction]}
                                type="primary"
                              >
                                <Wave>
                                  <button
                                    className="ant-btn ant-btn-primary"
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    <span
                                      key=".0"
                                    >
                                      OK
                                    </span>
                                  </button>
                                </Wave>
                              </Button>
                            </div>
                          </div>
                          <div
                            style={
                              Object {
                                "height": 0,
                                "overflow": "hidden",
                                "width": 0,
                              }
                            }
                            tabIndex={0}
                          >
                            sentinelEnd
                          </div>
                        </div>
                      </LazyRenderBox>
                    </AnimateChild>
                  </Animate>
                </div>
              </div>
            </Dialog>
          </Portal>
        </Portal>
      </DialogWrap>
    </Modal>
  </div>
</Details>
`;
